### PR TITLE
Improve performance of root slot in slot hashes history check

### PR DESCRIPTION
**Pull request by AshwinSekar**
#### Problem
Right now we do 2 iterations through slot hashes, one to check the root and one for the slots in the update.

#### Summary of Changes
Combine these into one iteration, and add a test.

### CODE REVIEW

File: programs/vote/src/vote_state/mod.rs [hunk 1]
```diff
@@ -530,13 +530,21 @@ impl VoteState {
                             // Then filter it out
                             vote_state_update_indexes_to_filter.push(vote_state_update_index);
                         }
-                        vote_state_update_index += 1;
+                        if check_root.is_some() {
+                            check_root = None;
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:534]: This implies the root is smaller than the smallest SlotAncestor, which can't happen due to the check here right: https://github.com/solana-labs/solana/blob/c463ee29f0284aebbf05b0bcbf04e8f3b067b762/programs/vote/src/vote_state/mod.rs#L475. We can assert this case is not true.

File: programs/vote/src/vote_state/mod.rs [hunk 2]
```diff
@@ -4170,6 +4182,36 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_check_update_vote_state_root_on_different_fork() {
+        let slot_hashes = build_slot_hashes(vec![2, 4, 6, 8]);
+        let vote_state = build_vote_state(vec![2, 4, 6], &slot_hashes);
+
+        // Test with a `vote_state_update` where there:
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:4190]: nit: remove "there"

File: programs/vote/src/vote_state/mod.rs (hunk 1 above)
Comment (AshwinSekar) [reply programs/vote/src/vote_state/mod.rs:534]: It can happen if the vote state's current root is too old to be in slot hashes history itself.

Comment (AshwinSekar) [reply programs/vote/src/vote_state/mod.rs:534]: e.g `test_check_update_vote_state_older_than_history_root`

Comment (carllin) [reply programs/vote/src/vote_state/mod.rs:534]: ah I see, thanks. can we add an assert here then that `self.root_slot < earliest_slot_hash_in_history` and add a comment describing why this is true?

**Pull request merged**